llvm/lib/Frontend/HLSL/HLSLRootSignature.cpp

```diff
@@ -0,0 +1,154 @@
+//===- HLSLRootSignature.cpp - HLSL Root Signature helper objects ---------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file This file contains helpers for working with HLSL Root Signatures.
+///
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Frontend/HLSL/HLSLRootSignature.h"
+#include "llvm/ADT/bit.h"
+
+namespace llvm {
+namespace hlsl {
+namespace rootsig {
+
+static void dumpRegType(raw_ostream &OS, RegisterType Type) {
```

Comment:
I think this whole file becomes a bit simpler and cleaner if we define a bunch of `operator<<`s instead of dump functions. So here we could have

```c++
static raw_ostream &operator<<(raw_ostream &OS, const Register &Reg) {
  switch (Reg.Type) {
  case RegisterType::BReg:
    OS << "b";
    break;
  // ...
  }
  OS << Number;
  return OS;
```

and similarly for the rest of these static dump functions. Then the publicly visible `dump()` methods can be implemented as `OS << "Foo(" << thing1 << "bar = " << thing2 << ")"` etc

Reply:
+1